quic/s2n-quic-platform/src/socket/io/rx.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{message::Message, socket::ring::Consumer};
+use core::task::{Context, Poll};
+use s2n_quic_core::{
+    event,
+    inet::datagram,
+    io::rx,
+    path::{LocalAddress, MaxMtu},
+};
+
+/// Structure for receiving messages from consumer channels
+pub struct Rx<T: Message> {
+    channels: Vec<Consumer<T>>,
+    max_mtu: MaxMtu,
+    local_address: LocalAddress,
+}
+
+impl<T: Message> Rx<T> {
+    #[inline]
+    pub fn new(channels: Vec<Consumer<T>>, max_mtu: MaxMtu, local_address: LocalAddress) -> Self {
+        Self {
+            channels,
+            max_mtu,
+            local_address,
+        }
+    }
+}
+
+impl<T: Message> rx::Rx for Rx<T> {
+    type PathHandle = T::Handle;
+    type Queue = RxQueue<'static, T>;
+    type Error = ();
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        let mut is_any_ready = false;
+        let mut is_all_closed = true;
+
+        // try to acquire any messages we can from the set of channels
+        for channel in &mut self.channels {
+            match channel.poll_acquire(u32::MAX, cx) {
+                Poll::Ready(_) => {
+                    is_all_closed = false;
+                    is_any_ready = true;
+                }
+                Poll::Pending => {
+                    is_all_closed &= !channel.is_open();
```

Comment:
Yeah I had that originally but it ended up forcing the implementation to call `is_open` more than needed in some situations. By splitting it out, the caller is in control of when synchronization takes place. But I agree it isn't the safest API either, since you can forget to call it.

Reply:
Makes sense. Seems ok.